File: C_projects/reading_grade_analyzer/readability.c
#include <ctype.h>
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <math.h>

string reading_grade(string phrase);                        //declare function

int main(void)
{
    string phrase = get_string("Text: ");                   //get input
    printf("%s\n", reading_grade(phrase));                  //print answer using function
}

string reading_grade(string phrase)                         //function to calculate and compute grade level for this assignment
{
    string ans = "";                                        //placeholder to return a string
    int words = 0, sentences = 0, letters = 0;
    for (int i = 0, n = strlen(phrase); i < n; i++)         //go through each char in string array
    {
        //Word = letter followed by space = word
        //OR
        //Word = letter OR punctuation followed by \0, i.e end of string
        if ((phrase[i] == ' ') || ((isalpha(phrase[i]) || ispunct(phrase[i])) && phrase[i + 1] == '\0'))
        {
            words++;
        }
        if (phrase[i] == '.' || phrase[i] == '!' || phrase[i] == '?') //check for punctuation for sentence count
        {
            sentences++;
        }
        if (isalpha(phrase[i]))                             //check for chars for letter count
        {
            letters++;
        }
    }
    float avgS  = (float) sentences / words * 100;              //compute avg sentences per 100 words
    float avgL  = (float) letters / words * 100;                //compute avg letters per 100 words
    int index   = round(0.0588 * avgL - 0.296 * avgS - 15.8);   //Coleman-Liau index
    if (index >= 16)
    {
        printf("Grade 16+");
    }
    else if (index < 1)
    {
        printf("Before Grade 1");
    }
    else
    {
        printf("Grade %i", index);
    }
    return ans;
}

File: Python_projects/rewrite_reading_grade_analyzer/readability.py
# get user input
text = input("Text: ")

# declare tracking variables
# words always starts at 1 to account for last word
# assumes user inputs atleast 1 word and not empty
letters = sentences = 0
words = 1

# point at each letter in text
for pointer in text:
    # if the pointer is an alphanumeric, increment "letters"
    if pointer.isalpha():
        letters += 1
    # if the pointer can be found in list (space), increment "words"
    if pointer in [' ']:
        words += 1
    # if the pointer can be found in list (./?/!), increment "sentences"
    if pointer in ['.', '?', '!']:
        sentences += 1

# L = the average number of letters per 100 words in the text
# S = the average number of sentences per 100 words in the text.
L = letters / words * 100
S = sentences / words * 100
index = 0.0588 * L - 0.296 * S - 15.8
grade = int(round(index))

if grade < 1:
    print("Before Grade 1")
elif grade > 16:
    print("Grade 16+")
else:
    print(f"Grade {grade}")